### PR TITLE
[16.07] Fix tool dependencies handling in Tool Shed model

### PR DESCRIPTION
**Pull request by davebx**

### CODE REVIEW

**Comment by nsoranzo** [comment]
Doesn't seem to fix it :sob:

**Comment by davebx** [comment]
The test will continue to fail, since it's trying to retrieve data from MTS, which does not have the updated code.

**Comment by martenson** [comment]
We need to port the fix to 16.04 too I think

File: lib/galaxy/webapps/tool_shed/controllers/repository.py [hunk 1]
```diff
@@ -1466,7 +1466,7 @@ def get_tool_dependencies( self, trans, **kwd ):
         changeset_revision = kwd.get( 'changeset_revision', None )
         repository = repository_util.get_repository_by_name_and_owner( trans.app, name, owner )
         dependencies = repository.get_tool_dependencies( changeset_revision )
-        if dependencies is not None:
+        if len( dependencies ) <= 0:
```
Comment (nsoranzo) [inline lib/galaxy/webapps/tool_shed/controllers/repository.py]: Should this be `>=` instead of `<=` ?
Or even just `if dependencies:`

**Comment by davebx** [comment]
@martenson Yeah, we probably should.

File: lib/galaxy/webapps/tool_shed/controllers/repository.py (hunk 1 above)
Comment (davebx) [reply lib/galaxy/webapps/tool_shed/controllers/repository.py]: I'm not a fan of `if <something that is neither explicitly True nor False>:`, and `<= 0` evaluates to True on an empty dict, which is what I'm looking for.

File: lib/galaxy/webapps/tool_shed/model/__init__.py [hunk 2]
```diff
@@ -246,10 +246,12 @@ def get_type_class( self, app ):
         return app.repository_types_registry.get_class_by_label( self.type )
 
     def get_tool_dependencies( self, changeset_revision ):
+        repo = hg.repository( ui.ui(), self.repo_path( app ) )
```
Comment (nsoranzo) [inline lib/galaxy/webapps/tool_shed/model/__init__.py:249]: `app` is not defined in this context, should be a function parameter?

See https://jenkins.galaxyproject.org/job/docker-toolshed/4129/testReport/junit/shed_functional.functional.test_0540_get_all_metadata_from_api/TestGetAllMetadata/test_0025_verify_dependency_json/

**Review by nsoranzo — CHANGES_REQUESTED** [review]

File: lib/galaxy/webapps/tool_shed/model/__init__.py [hunk 3]
```diff
@@ -246,10 +246,12 @@ def get_type_class( self, app ):
...
+        changeset_revision = suc.get_next_downloadable_changeset_revision( self, repo, changeset_revision )
```
Comment (nsoranzo) [inline lib/galaxy/webapps/tool_shed/model/__init__.py]: `get_next_downloadable_changeset_revision` is in `suc` in 16.04, but is in `metadata_util` in 16.07 and following releases. So you should retarget the PR to 16.04 or change this to `metadata_util`.

**Review by nsoranzo — CHANGES_REQUESTED** [review]

File: lib/galaxy/webapps/tool_shed/model/__init__.py (hunk 2 above)
Comment (nsoranzo) [reply lib/galaxy/webapps/tool_shed/model/__init__.py:249]: Also, `repo = hg_util.get_repo_for_repository( app, repository=self )` is probably cleaner, but would require to add `hg_util` to `from tool_shed.util import metadata_util`

File: lib/galaxy/webapps/tool_shed/api/repositories.py [hunk 4]
```diff
@@ -781,7 +781,7 @@ def metadata( self, trans, id, **kwd ):
             else:
                 metadata_dict[ 'repository_dependencies' ] = []
             if metadata.includes_tool_dependencies and recursive:
-                metadata_dict[ 'tool_dependencies' ] = repository.get_tool_dependencies( changehash )
+                metadata_dict[ 'tool_dependencies' ] = repository.get_tool_dependencies( trans.app, changehash )
```
Comment (nsoranzo) [inline lib/galaxy/webapps/tool_shed/api/repositories.py]: self.app ?

**Review by nsoranzo — APPROVED** [review]

**Pull request merged**